### PR TITLE
Fix windowing over time bugs

### PR DESCRIPTION
**Pull request by chinmaychandak**
This PR fixes two windowing bugs:

1. Incorrect output when intermediate indexes in windowing-over-time operations are empty: #357  

2. Make windowing-over-time output conform with Pandas windowing-over-time. E.g. The code below give output as 10, but the current output is 15.

```
from streamz import Stream
from streamz.dataframe import DataFrame
import pandas as pd

stream = Stream()

example = pd.DataFrame({"amount":[]})
sdf = DataFrame(stream=stream, example=example)

sdf.window("2h").sum().stream.sink(print)

stream.emit(pd.DataFrame({"amount":[5,5,5]}, index=[pd.Timestamp("2050-01-01 00:00:00"),
                                                    pd.Timestamp("2050-01-01 01:00:00"),
                                                    pd.Timestamp("2050-01-01 02:00:00")]))
```

### CODE REVIEW

File: streamz/dataframe/aggregations.py [hunk 1]
```diff
@@ -167,18 +167,20 @@ def diff_iloc(dfs, new, window=None):
         List of dataframes to decay
     """
     dfs = deque(dfs)
-    dfs.append(new)
+    if len(new) > 0:
```
Comment (martindurant) [inline streamz/dataframe/aggregations.py:170]: ~`not new.empty`~

^ it turns out that using `len` is consistently faster than using `empty`, I can't imagine why.

File: streamz/dataframe/aggregations.py [hunk 2]
```diff
@@ -202,16 +204,18 @@ def diff_loc(dfs, new, window=None):
...
-        old.append(o)  # TODO: avoid copy if fully lost
-        dfs[0] = dfs[0].iloc[len(o):]
-        if not len(dfs[0]):
-            dfs.popleft()
+    if len(dfs) > 0:
+        mx = max(df.index.max() for df in dfs)
+        mn = mx - pd.Timedelta(window) + pd.Timedelta('1ns')
+        while pd.Timestamp(dfs[0].index.min()) < mn:
+            o = dfs[0].loc[:mn]
+            old.append(o)  # TODO: avoid copy if fully lost
```
Comment (martindurant) [inline streamz/dataframe/aggregations.py:215]: May as well check for an empty `old` here and fix the TODO?

Comment (martindurant) [reply streamz/dataframe/aggregations.py:215]: `old = old.append(o) if len(old) else o`. (+ comment to explain)

File: streamz/dataframe/tests/test_dataframes.py [hunk 3]
```diff
@@ -755,12 +755,11 @@ def f(x):
...
 
     first = df.iloc[:diff]
-    lost = first[first.index.min() + value:]
-    first = first.iloc[len(lost):]
+    first = first[first.index.max() - value + pd.Timedelta('1ns'):]
 
     assert_eq(L[0], f(first))
 
-    last = df.loc[index.max() - value + pd.Timedelta('1s'):]
+    last = df.loc[index.max() - value + pd.Timedelta('1ns'):]
```
Comment (martindurant) [inline streamz/dataframe/tests/test_dataframes.py:762]: Should include your specific test case in the suite?

Comment (chinmaychandak) [reply streamz/dataframe/tests/test_dataframes.py:762]: Done

File: streamz/dataframe/aggregations.py (hunk 2 above)
Comment (chinmaychandak) [reply streamz/dataframe/aggregations.py]: Done

**Comment by chinmaychandak** [comment]
@martindurant Does this look good now?

**Comment by martindurant** [comment]
OK, accepting this. I feel like we could still use more testing here, there are probably more edges.

**Comment by chinmaychandak** [comment]
@martindurant, can we cut a release, please? Because these are pretty major windowing-over-time bugs we're seeing in our production pipelines.   

Sorry for the continual tagging in advance!

**Comment by martindurant** [comment]
0.5.6 now on pypi

**Comment by chinmaychandak** [comment]
Just curious, does the regro-cf-autotick-bot take care of the conda release automatically? I see that the previous releases were handled by the conda-forge streamz-feedstock.

**Comment by martindurant** [comment]
Yes: https://github.com/conda-forge/streamz-feedstock/pull/13

**Pull request merged**